audio/src/main.zig

```diff
@@ -0,0 +1,120 @@
+const std = @import("std");
+const mem = std.mem;
+const testing = std.testing;
+const builtin = @import("builtin");
+const Backend = switch (builtin.os.tag) {
+    .linux,
+    .windows,
+    .macos,
+    .ios,
+    => @import("soundio.zig"),
+    else => @compileError("unsupported os"),
+};
+pub const Error = Backend.Error;
+pub const Device = Backend.Device;
+pub const DeviceIterator = Backend.DeviceIterator;
+const DeviceDescriptorInternal = Backend.DeviceDescriptorInternal;
+
+pub const Mode = enum {
+    input,
+    output,
+};
+
+pub const Format = enum {
+    U8,
+    S16,
+    S24,
+    S32,
+    F32,
+};
+
+pub const DeviceDescriptor = struct {
+    mode: ?Mode = null,
+    format: ?Format = null,
+    is_raw: ?bool = null,
+    channels: ?u8 = null,
+    sample_rate: ?u32 = null,
+    id: ?[]const u8 = null,
+    name: ?[]const u8 = null,
+    callback: ?fn (device: Device, frame_count: u32) void = null,
+
+    internal: ?DeviceDescriptorInternal = null,
+};
+
+const Audio = @This();
+
+backend: Backend,
+
+pub fn init(allocator: mem.Allocator) Error!Audio {
+    return Audio{
+        .backend = try Backend.init(allocator),
+    };
+}
+
+pub fn deinit(self: Audio) void {
+    self.backend.deinit();
+}
+
+pub fn waitEvents(self: Audio) void {
+    self.backend.waitEvents();
+}
+
+pub fn requestDevice(self: Audio, config: DeviceDescriptor) Error!Device {
+    return self.backend.requestDevice(config);
+}
+
+pub fn inputDeviceIterator(self: Audio) DeviceIterator {
+    return self.backend.inputDeviceIterator();
+}
+
+pub fn outputDeviceIterator(self: Audio) DeviceIterator {
+    return self.backend.outputDeviceIterator();
+}
+
+test "list devices" {
+    const a = try init(testing.allocator);
+    defer a.deinit();
+
+    var iter = a.inputDeviceIterator();
+    while (try iter.next()) |d| {
+        if (d.id) |id|
+            testing.allocator.free(id);
```

Comment:
It's a bummer we need to free each ID, though, I don't like that much

Reply:
relax. this happens once game load

Reply:
How do you know the user would do it only once? Well you cant say it surely.

Reply:
it's user fault and beyond of our control. user can save to an array and use it where want

Reply:
User cannot really store it since devices may change as the program is running, which is triggered by ``flushEvents()`` in my understanding

Reply:
hmm right. but this seems the only way to keep id standalone. im almost ok with making it user job to save `is_raw` and `mode` itself but it seems other backends provide a unique id.